core/arch/arm/plat-zynqmp/drivers/aes.c

```diff
@@ -0,0 +1,295 @@
+	ret = csu_dma_sync(CSU_DMA_DST_CHANNEL);
+	if (ret) {
+		EMSG("DMA destination transfer timeout");
+		return ret;
+	}
+
+	csu_dma_unprepare();
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result aes_transfer_dec(uint8_t *src, uint8_t *dst, uint8_t *tag,
+				   uint8_t *iv, size_t len)
+{
+	TEE_Result ret = TEE_SUCCESS;
+
+	csu_dma_prepare();
+
+	/* Prepare destination */
+	csu_dma_transfer(CSU_DMA_DST_CHANNEL, dst, len / 4, 0);
+
+	/* Inputs */
+	csu_dma_transfer(CSU_DMA_SRC_CHANNEL, iv, GCM_TAG_SIZE / 4, 0);
+	ret = csu_dma_sync(CSU_DMA_SRC_CHANNEL);
+	if (ret) {
+		EMSG("DMA IV transfer timeout");
+		return ret;
+	}
+	csu_dma_transfer(CSU_DMA_SRC_CHANNEL, src, len / 4, CSU_DMA_DONE);
+	ret = csu_dma_sync(CSU_DMA_SRC_CHANNEL);
+	if (ret) {
+		EMSG("DMA source transfer timeout");
+		return ret;
+	}
+	csu_dma_transfer(CSU_DMA_SRC_CHANNEL, tag, GCM_TAG_SIZE / 4,
+			 CSU_DMA_DONE);
+	ret = csu_dma_sync(CSU_DMA_SRC_CHANNEL);
+	if (ret) {
+		EMSG("DMA tag transfer timeout");
+		return ret;
+	}
+
+	/* Wait for completion*/
+	ret = csu_dma_sync(CSU_DMA_DST_CHANNEL);
+	if (ret) {
+		EMSG("DMA destination transfer timeout");
+		return ret;
+	}
+
+	csu_dma_unprepare();
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result aes_prepare_op(enum aes_op op, enum csu_key key)
+{
+	vaddr_t aes = core_mmu_get_va(CSU_AES_BASE, MEM_AREA_IO_SEC,
+				      CSU_AES_SIZE);
+	vaddr_t csu = core_mmu_get_va(CSU_BASE, MEM_AREA_IO_SEC,
+				      0x1000);
+	TEE_Result ret = TEE_SUCCESS;
+
+	/* Connect DMA0 in/out to AES */
+	io_write32(csu + SSS_OFFSET, DMA0_STREAM_TO_AES);
+
+	/* Reset the AES */
+	io_write32(aes + AES_RESET_OFFSET, AES_RESET_SET);
+	io_write32(aes + AES_RESET_OFFSET, AES_RESET_CLR);
+
+	/* Load the key */
+	io_write32(aes + AES_KEY_CLR_OFFSET, 0);
+	io_write32(aes + AES_KEY_SRC_OFFSET, key);
+	io_write32(aes + AES_KEY_LOAD_OFFSET, AES_KEY_LOAD);
+	ret = aes_wait(AES_STS_KEY_INIT_DONE, true);
+	if (ret) {
+		EMSG("Timeout loading the key");
+		return TEE_ERROR_GENERIC;
+	}
+
+	/* Configure operation */
+	io_write32(aes + AES_CFG_OFFSET,
+		   op == AES_DEC ? AES_CFG_DEC : AES_CFG_ENC);
+
+	/* Prepare the CSU for the DMA */
+	io_write32(csu + DMA_RESET_OFFSET, DMA_RESET_SET);
+	io_write32(csu + DMA_RESET_OFFSET, DMA_RESET_CLR);
+
+	/* Start the message */
+	io_write32(aes + AES_START_MSG_OFFSET, AES_START_MSG);
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result aes_done_op(enum aes_op op, TEE_Result ret)
+{
+	vaddr_t aes = core_mmu_get_va(CSU_AES_BASE, MEM_AREA_IO_SEC,
+				      CSU_AES_SIZE);
+	uint32_t val = 0;
+
+	/* Validate the GCM tag */
+	if (!ret && op == AES_DEC) {
+		val = io_read32(aes + AES_STS_OFFSET) & AES_STS_GCM_TAG_OK;
+		if (!val) {
+			EMSG("AES-GCM tag mismatch");
+			return TEE_ERROR_GENERIC;
+		}
+	}
+
+	val = io_read32(aes + AES_KEY_CLR_OFFSET);
+	io_write32(aes + AES_KEY_CLR_OFFSET, val | AES_KEY_ZERO | AES_KUP_ZERO);
+	if (aes_wait(AES_STS_AES_KEY_ZEROED | AES_STS_KUP_ZEROED, true))
+		EMSG("Failed to clear the AES key");
```

Comment:
Should the function proceed even in case of error?

Reply:
I dont see why not: the encoding/decoding did complete, it is just the housekeeping...I suppose it might lead to a failure on the next run ...hence why I am doing the EMSG.

Reply:
ok